### PR TITLE
Add Prilliman et al transience model to pvlib.temperature

### LINKED ISSUE #1081: implement transient moving-average thermal model
**New Feature**
Implement the [transient moving-average model proposed by Matt Prillman](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9095219)

**Describe the solution you'd like**
a new function `pvlib.temperature.prillman(times, ...)`

**Describe alternatives you've considered**
see also #1080 

**Additional context**
see https://github.com/NREL/ssc/issues/261

### PR DESCRIPTION
**Pull request by kanderso-nrel**
- [x] Closes #1081
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

This PR adds a cleaned-up version of `prilliman_v5` from this notebook: https://gist.github.com/kanderso-nrel/1d6da384d7af8afc24c230f1f144eb57

Also: I noticed that one of the links in the PR template is out of date so I updated that here too.

### CODE REVIEW

**Comment by kanderso-nrel** [comment]
Ready for review.  

@mjprilliman comments are welcome :)

**Review by cwhanse — COMMENTED** [review]
Haven't looked at the test details yet

File: pvlib/temperature.py
Comment (adriesse) [inline pvlib/temperature.py:907]: If `timestep` is in seconds, then you would have to divide by 60 to get minutes...

Comment (kanderso-nrel) [reply pvlib/temperature.py:907]:
```suggestion
    timedeltas = np.arange(window, 0, -1) * (time_step*60)  # min to s
```

**Comment by wholmgren** [comment]
Great to see this. I'll review soon.

Reproducing something like Figure 5 from the reference would be a great addition to the example library.

prilliman figure 5 winter

File: pvlib/temperature.py
Comment (wholmgren) [inline pvlib/temperature.py:909]: The array manipulation below is very clever, but will fly over most people's heads (well, at least I had to look up the function examples). I don't think the code needs to change, and I don't have a great idea for how to make it more understandable, and it requires understanding the hankel matrix too, nor do I think it's your job to teach numpy here, but any further comments below might help. So yeah, tough spot, sorry.

File: pvlib/temperature.py
Comment (wholmgren) [inline pvlib/temperature.py:893]: I like this comment in `clearsky.py` "# generate matrix of integers for creating windows with indexing"

File: pvlib/temperature.py
Comment (wholmgren) [inline pvlib/temperature.py:921]: why nansum above but sum here?

**Review by wholmgren — COMMENTED** [review]
Looks good! All minor (hopefully) comments below.

**Review by mjprilliman — COMMENTED** [review]
I'm not familiar enough with pvlib to give a full review but the function implementation looks good to me. Thanks @kanderso-nrel !

File: pvlib/temperature.py
Comment (kanderso-nrel) [reply pvlib/temperature.py:886]: Indeed I took that behavior from @mjprilliman's example code in #1081, but I think using whatever is available is also consistent with Equation 11 in the reference.  There is some relevant discussion in this comment and the ones following: https://github.com/pvlib/pvlib-python/issues/1081#issuecomment-1019593000

Either way I am +1 to adding a docstring note about the partial windows at the beginning :+1:

File: pvlib/temperature.py
Comment (cwhanse) [reply pvlib/temperature.py:834]: I'm -1 on using the term "exponentially weighted" because that implies that the weight are of the form exp(k * delta_t) where k is a constant. In this model, k=k(t) is a function of t (P in Eq. 11 changes at each time step).

File: pvlib/tests/test_temperature.py
Comment (kanderso-nrel) [reply pvlib/tests/test_temperature.py:344]: My idea was that each calculation operates on a single value of wind speed and many values of temperature, so a nan wind speed is a deal breaker but a nan temperature is ok so long as there are some non-nan temperatures in the window.  The reference indicates that gaps in the data are acceptable, which I think is functionally equivalent to ignoring nans.  I'm open to switching to nan output if any of the input temperature is nan, but to me that seems less consistent with the reference.

Comment (wholmgren) [reply pvlib/tests/test_temperature.py:344]: Thanks, now I agree with you. An `Examples` section of the docstring might help prevent user confusion, but I don't want to require that for this PR.

**Comment by kanderso-nrel** [comment]
Thanks @wholmgren for the comments, was just lazy of me to not document that numpy magic.  I came up with it and still had to remind myself how it worked!

**Review by wholmgren — APPROVED** [review]
Any further comments or should we merge?

**Comment by cwhanse** [comment]
I'd like to take one more look.

File: pvlib/temperature.py
Comment (cwhanse) [inline pvlib/temperature.py:870]:
```suggestion
    Smoothing is done using the 20 minute window behind each temperature
    value. At the beginning of the series where a full 20 minute window is not
    possible, partial windows are used instead.
```

**Review by cwhanse — APPROVED** [review]
You win some unnamed award for excellent in-line comments

**Comment by kanderso-nrel** [comment]
Actually I think there may be a flaw in this implementation regarding treatment of nans.  Please do not merge yet!

**Comment by kanderso-nrel** [comment]
Ok, fixed.  The problem was that NaNs in the input temperature were correctly ignored in the numerator (`np.nansum(subsets * weights, axis=1)`), but the corresponding weights were not set to zero in the denominator, so in effect it was treating NaN temperature as zero temperature instead of ignoring it.  The fix was to replace the `triu` incantation, which properly ignored NaNs for the "prefix", with a simple mask that ignores all NaNs, not just the prefix.  So the code is simpler now too.

I don't think this should be controversial but it's probably worth another pair of eyes.  Thanks!

File: pvlib/temperature.py
Comment (cwhanse) [inline pvlib/temperature.py:872]: One more suggestion: explain in Notes how `nan` in inputs corresponds to `nan` in output. It makes sense (to me) that output at `k` is `nan` if `windspeed[k]` is `nan`. As a user, I may be surprised non-`nan` at position `k` in output when input `temp_cell[k]` is `nan`. But that also makes sense, because the non-`nan` value can be produced when window of `temp_cell` prior to `k` has data.

```
Output ``temp_cell[k]`` is NaN when input ``wind_speed[k]`` is NaN, or when no non-NaN data are
in the input temperature for the 20 minute window preceding index ``k``.
```

**Comment by adriesse** [comment]
> You win some unnamed award for excellent in-line comments

I had noticed these also, with appreciation.

**Comment by cwhanse** [comment]
If not objections, will merge tomorrow.

File: pvlib/temperature.py
Comment (wholmgren) [inline pvlib/temperature.py:976]: @kanderso-nrel this line emitted a warning despite the comment on lines 967-969 that suggested otherwise. Is this a version-specific issue? My versions are below...

```
INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.9.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18363
machine          : AMD64
processor        : Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.3.5
numpy            : 1.20.3
```

Comment (kanderso-nrel) [reply pvlib/temperature.py:976]: Hmm, the only way I can reproduce this is if there is a stretch of NaN in the input lasting longer than 20 minutes -- is that true for your case?  That comment is referring to a div by zero warning cause by a quirk of this implementation (no weights for the very first value), not a div by zero warning caused by nans in the input.

Comment (wholmgren) [reply pvlib/temperature.py:976]: Ah, thanks. Yes, I had 38 minutes of NaN temperatures. I filled only the middle point with a valid number and the warning disappeared.

**Pull request merged**
